File: nexus-backend/src/main/java/com/nexusop/shared/util/AmountInWords.java
package com.nexusop.shared.util;

import java.math.BigDecimal;

/**
 * Converts numeric amounts to Indian English words.
 * Uses Indian number system: Crore, Lakh, Thousand.
 * 
 * Example: 873200 → "Eight Lakh Seventy Three Thousand Two Hundred Rupees Only"
 * Example: 15234567.50 → "One Crore Fifty Two Lakh Thirty Four Thousand Five Hundred Sixty Seven Rupees and Fifty Paise Only"
 */
public class AmountInWords {

    private static final String[] ONES = {
        "", "One", "Two", "Three", "Four", "Five", "Six", "Seven",
        "Eight", "Nine", "Ten", "Eleven", "Twelve", "Thirteen",
        "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"
    };

    private static final String[] TENS = {
        "", "", "Twenty", "Thirty", "Forty", "Fifty",
        "Sixty", "Seventy", "Eighty", "Ninety"
    };

    private static String twoDigits(int n) {
        if (n < 20) return ONES[n];
        String result = TENS[n / 10];
        if (n % 10 != 0) result += " " + ONES[n % 10];
        return result;
    }

    private static String threeDigits(int n) {
        if (n >= 100) {
            String result = ONES[n / 100] + " Hundred";
            if (n % 100 != 0) result += " " + twoDigits(n % 100);
            return result;
        }
        return twoDigits(n);
    }

    /**
     * Convert a BigDecimal amount to Indian English words.
     * @param amount The monetary amount (e.g., 873200.00)
     * @return "Eight Lakh Seventy Three Thousand Two Hundred Rupees Only"
     */
    public static String convert(BigDecimal amount) {
        if (amount == null || amount.compareTo(BigDecimal.ZERO) == 0) {
            return "Zero Rupees Only";
        }

        boolean negative = amount.compareTo(BigDecimal.ZERO) < 0;
        if (negative) amount = amount.abs();

        long rupees = amount.longValue();
        int paise = amount.subtract(BigDecimal.valueOf(rupees))
                         .multiply(BigDecimal.valueOf(100))
                         .intValue();

        StringBuilder result = new StringBuilder();
        if (negative) result.append("Minus ");

        // Indian number system: Crore → Lakh → Thousand → Hundred
        int crore = (int) (rupees / 10_000_000);
        int lakh  = (int) ((rupees % 10_000_000) / 100_000);
        int thou  = (int) ((rupees % 100_000) / 1_000);
        int rem   = (int) (rupees % 1_000);

        if (crore > 0) result.append(threeDigits(crore)).append(" Crore ");
        if (lakh > 0)  result.append(threeDigits(lakh)).append(" Lakh ");
        if (thou > 0)  result.append(threeDigits(thou)).append(" Thousand ");
        if (rem > 0)   result.append(threeDigits(rem));

        String rupeePart = result.toString().trim();
        if (rupeePart.isEmpty()) rupeePart = "Zero";
        rupeePart += " Rupees";

        if (paise > 0) {
            rupeePart += " and " + twoDigits(paise) + " Paise";
        }

        return rupeePart + " Only";
    }

    /**
     * Convenience method for double values.
     */
    public static String convert(double amount) {
        return convert(BigDecimal.valueOf(amount));
    }
}

File: backend/shared/amountInWords.js
/* ══════════════════════════════════════════════════════════
   amountInWords — Indian numbering (lakh, crore) for documents.

   This exact function was copy-pasted into five controllers. Consolidated
   here after createOrder referenced it and crashed, because the copy it
   needed happened to live in a different file. Five copies of a rounding
   rule is five chances for two documents to spell the same amount
   differently.
   ══════════════════════════════════════════════════════════ */
const ONES = ['', 'One', 'Two', 'Three', 'Four', 'Five', 'Six', 'Seven', 'Eight', 'Nine', 'Ten',
  'Eleven', 'Twelve', 'Thirteen', 'Fourteen', 'Fifteen', 'Sixteen', 'Seventeen', 'Eighteen', 'Nineteen'];
const TENS = ['', '', 'Twenty', 'Thirty', 'Forty', 'Fifty', 'Sixty', 'Seventy', 'Eighty', 'Ninety'];

const two = (n) => (n < 20 ? ONES[n] : TENS[Math.floor(n / 10)] + (n % 10 ? ' ' + ONES[n % 10] : ''));
const three = (n) =>
  (Math.floor(n / 100) ? ONES[Math.floor(n / 100)] + ' Hundred' + (n % 100 ? ' ' : '') : '') +
  (n % 100 ? two(n % 100) : '');

function amountInWords(value) {
  let num = Math.round(Number(value) || 0);
  if (num === 0) return 'Rupees Zero Only';
  const negative = num < 0;
  num = Math.abs(num);

  let out = '';
  const crore = Math.floor(num / 10000000); num %= 10000000;
  const lakh = Math.floor(num / 100000);    num %= 100000;
  const thousand = Math.floor(num / 1000);  num %= 1000;

  if (crore) out += three(crore) + ' Crore ';
  if (lakh) out += two(lakh) + ' Lakh ';
  if (thousand) out += two(thousand) + ' Thousand ';
  if (num) out += three(num);

  return `Rupees ${negative ? 'Minus ' : ''}${out.trim().replace(/\s+/g, ' ')} Only`;
}

module.exports = { amountInWords };
